File: speech_to_text/go/async/local_file/local_file.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"mime/multipart"
	"net/http"
	"os"
	"time"
)

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
var apiKey = os.Getenv("SONIOX_API_KEY")
var apiBase = "https://api.soniox.com"
var fileToTranscribe = "coffee_shop.mp3"

func pollUntilComplete(transcriptionId string) {
	for {
		req, err := http.NewRequest("GET", fmt.Sprintf("%s/v1/transcriptions/%s", apiBase, transcriptionId), nil)
		if err != nil {
			panic(err)
		}
		req.Header.Set("Authorization", "Bearer "+apiKey)
		res, err := http.DefaultClient.Do(req)
		if err != nil {
			panic(err)
		}
		defer res.Body.Close()
		if res.StatusCode != http.StatusOK {
			panic(res.Status)
		}
		var statusResponse map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&statusResponse); err != nil {
			panic(err)
		}
		if status, ok := statusResponse["status"].(string); ok {
			if status == "error" {
				panic("Transcription error: " + statusResponse["error_message"].(string))
			} else if status == "completed" {
				break
			}
		}
		time.Sleep(1 * time.Second)
	}
}

func main() {
	fmt.Println("Starting file upload...")

	fileBytes, err := os.ReadFile(fileToTranscribe)
	if err != nil {
		panic(err)
	}
	var fileUploadBody bytes.Buffer
	fileUploadWriter := multipart.NewWriter(&fileUploadBody)
	fileUploadWriterPart, err := fileUploadWriter.CreateFormFile("file", fileToTranscribe)
	if err != nil {
		panic(err)
	}
	fileUploadWriterPart.Write(fileBytes)
	fileUploadWriter.Close()
	req, err := http.NewRequest("POST", fmt.Sprintf("%s/v1/files", apiBase), &fileUploadBody)
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", fileUploadWriter.FormDataContentType())
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusCreated {
		panic(res.Status)
	}
	var file map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&file); err != nil {
		panic(err)
	}
	fileId := file["id"].(string)

	fmt.Println("Starting transcription...")

	transcriptionPayload := map[string]any{
		"file_id":        fileId,
		"model":          "stt-async-preview",
		"language_hints": []string{"en", "es"},
	}
	transcriptionBody, err := json.Marshal(transcriptionPayload)
	if err != nil {
		panic(err)
	}
	req, err = http.NewRequest("POST", fmt.Sprintf("%s/v1/transcriptions", apiBase), bytes.NewReader(transcriptionBody))
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	res, err = http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusCreated {
		panic(res.Status)
	}
	var transcription map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&transcription); err != nil {
		panic(err)
	}
	transcriptionId := transcription["id"].(string)

	fmt.Println("Transcription ID: " + transcriptionId)

	pollUntilComplete(transcriptionId)

	// Get the transcript text
	req, err = http.NewRequest("GET", fmt.Sprintf("%s/v1/transcriptions/%s/transcript", apiBase, transcriptionId), nil)
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	res, err = http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		panic(res.Status)
	}
	var transcriptResponse map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&transcriptResponse); err != nil {
		panic(err)
	}
	fmt.Printf("Transcript:\n%s\n", transcriptResponse["text"].(string))

	// Delete the transcription
	req, err = http.NewRequest("DELETE", fmt.Sprintf("%s/v1/transcriptions/%s", apiBase, transcriptionId), nil)
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	res, err = http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusNoContent {
		panic(res.Status)
	}

	// Delete the file
	req, err = http.NewRequest("DELETE", fmt.Sprintf("%s/v1/files/%s", apiBase, fileId), nil)
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	res, err = http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusNoContent {
		panic(res.Status)
	}
}

File: speech_to_text/csharp/async/local_file/LocalFile.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

// Retrieve the API key from the environment variable
string? apiKey = Environment.GetEnvironmentVariable("SONIOX_API_KEY");
if (string.IsNullOrEmpty(apiKey))
{
    throw new InvalidOperationException("SONIOX_API_KEY is not set.");
}
string apiBase = "https://api.soniox.com";
string fileToTranscribe = "../../coffee_shop.mp3";

var client = new HttpClient();
client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

async Task pollUntilComplete(string transcriptionId) {
    while (true)
    {
        var res = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}");
        res.EnsureSuccessStatusCode();
        var resJson = await res.Content.ReadAsStringAsync();
        var transcription = JsonNode.Parse(resJson);
        var status = transcription?["status"]?.ToString();
        if (string.IsNullOrEmpty(status))
        {
            throw new InvalidOperationException("Status is missing in transcription response.");
        }
        if (status == "completed")
        {
            return;
        }
        else if (status == "error")
        {
            var errorMessage = transcription?["error_message"]?.ToString() ?? "Unknown error";
            throw new Exception($"Transcription error: {errorMessage}");
        }
        await Task.Delay(1000);
    }
}

Console.WriteLine("Starting file upload...");

using var fileStream = new FileStream(fileToTranscribe, FileMode.Open, FileAccess.Read);
var fileContent = new MultipartFormDataContent();
fileContent.Add(new StreamContent(fileStream), "file", Path.GetFileName(fileToTranscribe));
var res = await client.PostAsync($"{apiBase}/v1/files", fileContent);
res.EnsureSuccessStatusCode();
var resJson = await res.Content.ReadAsStringAsync();
var file = JsonNode.Parse(resJson);
var fileId = file?["id"]?.ToString();
if (string.IsNullOrEmpty(fileId))
{
    throw new InvalidOperationException("File ID is missing after upload.");
}

Console.WriteLine("Starting transcription...");

var request = new StringContent(
    JsonSerializer.Serialize(new {
        file_id = fileId,
        model = "stt-async-preview",
        language_hints = new []{"en", "es"}
    }),
    System.Text.Encoding.UTF8,
    "application/json"
);
res = await client.PostAsync($"{apiBase}/v1/transcriptions", request);
res.EnsureSuccessStatusCode();
resJson = await res.Content.ReadAsStringAsync();
var transcription = JsonNode.Parse(resJson);
var transcriptionId = transcription?["id"]?.ToString();
if (string.IsNullOrEmpty(transcriptionId))
{
    throw new InvalidOperationException("Transcription ID is missing.");
}

Console.WriteLine($"Transcription ID: {transcriptionId}");

await pollUntilComplete(transcriptionId);

// Get the transcript text
res = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}/transcript");
res.EnsureSuccessStatusCode();
resJson = await res.Content.ReadAsStringAsync();
var transcript = JsonNode.Parse(resJson);
var transcriptText = transcript?["text"]?.ToString();
if (string.IsNullOrEmpty(transcriptText))
{
    throw new InvalidOperationException("Transcript text is missing.");
}
Console.WriteLine("Transcript:");
Console.WriteLine(transcriptText);

// Delete the transcription
res = await client.DeleteAsync($"{apiBase}/v1/transcriptions/{transcriptionId}");
res.EnsureSuccessStatusCode();

// Delete the file
res = await client.DeleteAsync($"{apiBase}/v1/files/{fileId}");
res.EnsureSuccessStatusCode();
